hypothesis-python/src/hypothesis/extra/numpy.py

```diff
@@ -177,14 +177,23 @@ def do_draw(self, data):
         if 0 in self.shape:
             return np.zeros(dtype=self.dtype, shape=self.shape)
 
+        # Because Numpy allocates memory for strings at array creation, if we have
+        # an unsized string dtype we'll fill an object array and then cast it back.
+        unsized_string_dtype = (
+            self.dtype.kind in (u"S", u"a", u"U") and self.dtype.itemsize == 0
+        )
+
         # Reset this flag for each test case to emit warnings from set_element
         # Skip the check for object or void (multi-element) dtypes
-        self._report_overflow = self.dtype.kind not in ("O", "V")
+        self._report_overflow = not (unsized_string_dtype or self.dtype.kind in "OV")
```

Comment:
Sorry for the nit here, but I think the legibility of `not in` is too nice to pass up on 
```suggestion
        self._report_overflow = not unsized_string_dtype and self.dtype.kind not in "OV"
```